Issue 233 divide by zero on transitions

bigredfrog opened the pull request:
Should address https://github.com/LedFx/LedFx/issues/233

Two issues

Transitions are created for a transition with zero time, but the mechanism to clear transitions is protected from zero transition time in the config

Don't create a transition when it should be immeadiate

Secondly don't test the config during transition processing, only at transition creation

Otherwise, transitions can be left orphened as the config can change async from the active transition

bigredfrog commented on ledfx/virtuals.py:
```diff
@@ -395,8 +400,6 @@ def assemble_frame(self):
             self._transition_effect is not None
             and self._transition_effect.is_active
             and hasattr(self._transition_effect, "pixels")
-            and self._config["transition_mode"] != "None"
```
These are removed to prevent an in flight transition getting orphaned, as they reference _config, which can change async. Instead we never create a transition in these circumstances

bigredfrog commented on ledfx/virtuals.py:
```diff
@@ -267,17 +267,22 @@ def set_effect(self, effect):
-        if self._active_effect is None:
-            self._transition_effect = DummyEffect(self.pixel_count)
+        if (
+            self._config["transition_mode"] != "None"
```
Only allow a transition creation if the current _config is enabled and non zero

bigredfrog commented on ledfx/virtuals.py:
```diff
@@ -267,17 +267,22 @@ def set_effect(self, effect):
+                self.clear_transition_effect()
+                self._transition_effect = self._active_effect
         else:
             self.clear_transition_effect()
```
speculative clear transition just in case, function protects against "not active"

YeonV commented:
Sooo nice brooo!!! @not-matt a python-god needs to approve it :)

not-matt commented:
Looks good to me! Just make sure it's well tested, especially when changing effects quickly and repeatedly.

bigredfrog commented:
Fast transitions is how I exposed it.

Tested by going up and down to zero transition time repeatedly while changing effects in between, had deterministic crash prior, now runs clean

Long transition time, trigger a transition and check the old transition does not get orphaned, added debug to watch for create / destroy on transitions, can see improved behaviour

Removed any race condition from the change of _config from the front end. Simplifies lifecycle, but no negative test

bigredfrog commented:
any comment on the linting failure, I don't think I can rerun it to see if it was a bounce

shauneccles commented:
Linting failure is due to flake8 moving from gitlab to github.

LGTM.
